caravel/viz.py

```diff
@@ -1660,6 +1660,174 @@ class HorizonViz(NVD3TimeSeriesViz):
+            'mapbox_color',
+        )
+    }, {
+        'label': 'Viewport',
+        'fields': (
+            'viewport_longitude',
+            'viewport_latitude',
+            'viewport_zoom',
+        )
+    },)
+
+    form_overrides = {
+        'all_columns_x': {
+            'label': 'Longitude',
+            'description': "Column containing longitude data",
+        },
+        'all_columns_y': {
+            'label': 'Latitude',
+            'description': "Column containing latitude data",
+        },
+        'pandas_aggfunc': {
+            'label': 'Cluster label aggregator',
+            'description': _(
+                "Aggregate function applied to the list of points "
+                "in each cluster to produce the cluster label."),
+        },
+        'rich_tooltip': {
+            'label': 'Tooltip',
+            'description': _(
+                "Show a tooltip when hovering over points and clusters "
+                "describing the label"),
+        },
+        'groupby': {
+            'description': _(
+                "One or many fields to group by. If grouping, latitude "
+                "and longitude columns must be present."),
+        },
+    }
+
+    def query_obj(self):
+        d = super(MapboxViz, self).query_obj()
+        fd = self.form_data
+        label_col = fd.get('mapbox_label')
+
+        if not fd.get('groupby'):
+            d['columns'] = [fd.get('all_columns_x'), fd.get('all_columns_y')]
+
+            if label_col and len(label_col) >= 1:
+                if label_col[0] == "count":
+                    raise Exception(
+                        "Must have a [Group By] column to have 'count' as the [Label]")
+                d['columns'].append(label_col[0])
+
+            if fd.get('point_radius') != 'Auto':
+                d['columns'].append(fd.get('point_radius'))
+
+            d['columns'] = list(set(d['columns']))
+        else:
+            # Ensuring columns chosen are all in group by
+            if (label_col and len(label_col) >= 1
+                and label_col[0] != "count"
+                and label_col[0] not in fd.get('groupby')):
+                raise Exception(
+                    "Choice of [Label] must be present in [Group By]")
+
+            if (fd.get("point_radius") != "Auto"
+                and fd.get("point_radius") not in fd.get('groupby')):
+                raise Exception(
+                    "Choice of [Point Radius] must be present in [Group By]")
+
+            if (fd.get('all_columns_x') not in fd.get('groupby') or
+                fd.get('all_columns_y') not in fd.get('groupby')):
+                raise Exception(
+                    "[Longitude] and [Latitude] columns must be present in [Group By]")
+        return d
+
+    def get_data(self):
+        df = self.get_df()
+        fd = self.form_data
+        label_col = fd.get('mapbox_label')
+        custom_metric = label_col and len(label_col) >= 1
+        metric_col = [None] * len(df.index)
```

Comment:
There's definitely a more panda-esque way to do all this, but it's not shocking as is.